samples/matter/common/src/bridge/ble_connectivity_manager.cpp

```diff
@@ -0,0 +1,310 @@
+	bt_gatt_dm_data_print(dm);
+	VerifyOrExit(bt_uuid_cmp(gatt_service->uuid, device->mServiceUuid) == 0, );
+	device->mConn = bt_gatt_dm_conn_get(dm);
+
+	discoveryResult = true;
+exit:
+
+	device->mConnectedCallback(device, dm, discoveryResult, device->mConnectedCallbackContext);
+}
+
+void BLEConnectivityManager::DiscoveryNotFound(struct bt_conn *conn, void *context)
+{
+	LOG_ERR("GATT service could not be found during the discovery");
+
+	BLEBridgedDevice *device = reinterpret_cast<BLEBridgedDevice *>(context);
+
+	device->mConnectedCallback(device, nullptr, false, device->mConnectedCallbackContext);
+}
+
+void BLEConnectivityManager::DiscoveryError(struct bt_conn *conn, int err, void *context)
+{
+	LOG_ERR("The GATT discovery procedure failed with %d", err);
+
+	BLEBridgedDevice *device = reinterpret_cast<BLEBridgedDevice *>(context);
+
+	device->mConnectedCallback(device, nullptr, false, device->mConnectedCallbackContext);
+}
+
+CHIP_ERROR BLEConnectivityManager::Init(bt_uuid **serviceUuids, uint8_t serviceUuidsCount)
+{
+	if (!serviceUuids || serviceUuidsCount > kMaxServiceUuids) {
+		return CHIP_ERROR_INVALID_ARGUMENT;
+	}
+
+	memcpy(mServicesUuid, serviceUuids, serviceUuidsCount * sizeof(serviceUuids));
+	mServicesUuidCount = serviceUuidsCount;
+
+	k_timer_init(&mScanTimer, BLEConnectivityManager::ScanTimeoutCallback, nullptr);
+	k_timer_user_data_set(&mScanTimer, this);
+
+	struct bt_scan_init_param scan_init = {
+		.connect_if_match = 0,
+	};
+
+	bt_scan_init(&scan_init);
+	bt_scan_cb_register(&scan_cb);
+
+	int err;
+
+	for (uint8_t i = 0; i < mServicesUuidCount; i++) {
+		err = bt_scan_filter_add(BT_SCAN_FILTER_TYPE_UUID, mServicesUuid[i]);
+
+		if (err) {
+			LOG_ERR("Failed to set scanning filter");
+			return System::MapErrorZephyr(err);
+		}
+	}
+
+	err = bt_scan_filter_enable(BT_SCAN_UUID_FILTER, false);
+	if (err) {
+		LOG_ERR("Filters cannot be turned on");
+		return System::MapErrorZephyr(err);
+	}
+
+	return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BLEConnectivityManager::Scan()
+{
+	if (mScanActive) {
+		LOG_ERR("Scan is already in progress");
+		return CHIP_ERROR_INCORRECT_STATE;
+	}
+
+	mScannedDevicesCounter = 0;
+
+	int err = bt_scan_start(BT_SCAN_TYPE_SCAN_ACTIVE);
+	if (err) {
+		LOG_ERR("Scanning failed to start (err %d)", err);
+		return System::MapErrorZephyr(err);
+	}
+
+	k_timer_start(&mScanTimer, K_MSEC(kScanTimeoutMs), K_NO_WAIT);
+	mScanActive = true;
+
+	return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BLEConnectivityManager::StopScan()
+{
+	if (!mScanActive) {
+		return CHIP_ERROR_INCORRECT_STATE;
+	}
+
+	int err = bt_scan_stop();
+	if (err) {
+		LOG_ERR("Scanning failed to stop (err %d)", err);
+		return System::MapErrorZephyr(err);
+	}
+
+	mScanActive = false;
+
+	return CHIP_NO_ERROR;
+}
+
+void BLEConnectivityManager::ScanTimeoutCallback(k_timer *timer)
+{
+	DeviceLayer::PlatformMgr().ScheduleWork(ScanTimeoutHandle, reinterpret_cast<intptr_t>(timer));
+}
+
+void BLEConnectivityManager::ScanTimeoutHandle(intptr_t context)
```

Comment:
I think that ideally we should let the application to register this kind of callback and decide what to do with scan data. I can imagine the real device not wanting to interact with the console at all.

We may also think about single scan-result per callback, to let the application filter the particular device and (for instance) abort the scan once the device of interest is found (it would require API extension).

Reply:
> I think that ideally we should let the application to register this kind of callback and decide what to do with scan data

Totally agree, I wanted to implement it, but then I forgot to do that.

> We may also think about single scan-result per callback

Unfortunately Bluetooth API is not as good as Wi-Fi in this matter. In Bluetooth you will receive the same result from the same address multiple times (e.g. if you have 3 devices you will receive like ~30 calls in 10 s, 10 per each device), as it just reports the fact every time it captures advertisement. That's why I decided to create a list in the Bluetooth module and pass processed data to the application.

Reply:
Ok, sounds fair.